File: app/src/main/java/starbright/com/projectegg/features/detail/RecipeDetailFragment.java
/**
 * Created by Andreas on 29/9/2018.
 */

package starbright.com.projectegg.features.detail;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.NestedScrollView;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.CardView;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import starbright.com.projectegg.MyApp;
import starbright.com.projectegg.R;
import starbright.com.projectegg.data.AppRepository;
import starbright.com.projectegg.data.local.model.Ingredient;
import starbright.com.projectegg.data.local.model.Instruction;
//import starbright.com.projectegg.util.GlideApp;
import starbright.com.projectegg.util.TextViewRecyclerAdapter;
import starbright.com.projectegg.util.scheduler.BaseSchedulerProvider;

public class RecipeDetailFragment extends Fragment implements RecipeDetailContract.View {

    private static final String BUNDLE_RECIPE_ID = "BUNDLE_RECIPE_ID";

    @BindView(R.id.img_banner_food)
    ImageView imgBannerFood;

    @BindView(R.id.container_header)
    RelativeLayout containerHeader;

    @BindView(R.id.card_ingredient)
    CardView cardIngredient;

    @BindView(R.id.card_instruction)
    CardView cardInstruction;

    @BindView(R.id.rv_ingredient)
    RecyclerView rvIngredient;

    @BindView(R.id.rv_instruction)
    RecyclerView rvInstruction;

    @BindView(R.id.tv_serving_count)
    TextView tvServingCount;

    @BindView(R.id.tv_preparation_time)
    TextView tvPreparationTime;

    @BindView(R.id.tv_cooking_time)
    TextView tvCookingTime;

    @BindView(R.id.scroll_container)
    NestedScrollView scrollContainer;

    @BindView(R.id.tv_empty_text)
    TextView tvEmptyText;

    @BindView(R.id.swipe_refresh_container)
    SwipeRefreshLayout swipeRefreshLayout;

    @BindView(R.id.tv_recipe_title)
    TextView tvRecipeName;

    @BindView(R.id.ad_view)
    AdView adView;

    @Inject
    AppRepository repository;

    @Inject
    BaseSchedulerProvider schedulerProvider;

    private RecipeDetailContract.Presenter mPresenter;
    private FragmentListener mFragmentListener;

    public static RecipeDetailFragment newInstance(String recipeId) {
        Bundle args = new Bundle();
        args.putString(BUNDLE_RECIPE_ID, recipeId);
        RecipeDetailFragment fragment = new RecipeDetailFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        MyApp.getAppComponent().inject(this);
        super.onAttach(context);
        mFragmentListener = (FragmentListener) context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        new RecipeDetailPresenter(repository, this, schedulerProvider);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_recipe_detail, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mPresenter.setRecipeId(getArguments().getString(BUNDLE_RECIPE_ID));
        mPresenter.start();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.recipe_detail_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_share:
                mPresenter.handleShareMenuClicked();
                return true;
            case R.id.menu_webview:
                mPresenter.handleWebViewMenuClicked();
                return true;
        }
        return false;
    }

    @Override
    public void setPresenter(RecipeDetailPresenter presenter) {
        mPresenter = presenter;
    }

    @Override
    public void showProgressBar() {
        swipeRefreshLayout.setRefreshing(true);
    }

    @Override
    public void hideProgressBar() {
        swipeRefreshLayout.setRefreshing(false);
    }

    @Override
    public void hideScrollContainer() {
        scrollContainer.setVisibility(View.GONE);
    }

    @Override
    public void showScrollContainer() {
        scrollContainer.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideEmptyStateTextView() {
        tvEmptyText.setVisibility(View.GONE);
    }

    @Override
    public void renderErrorStateTextView(String errorMessage) {
        tvEmptyText.setVisibility(View.VISIBLE);
        tvEmptyText.setText(errorMessage);
    }

    @Override
    public void renderEmptyStateTextView() {
        tvEmptyText.setVisibility(View.VISIBLE);
        tvEmptyText.setText(getString(R.string.detail_empty_label));
    }

    @Override
    public void renderBannerFoodImage(String imageUrl) {
        imgBannerFood.setVisibility(View.VISIBLE);
//        GlideApp.with(getActivity())
//                .load(imageUrl)
//                .diskCacheStrategy(DiskCacheStrategy.ALL)
//                .centerCrop()
//                .into(imgBannerFood);
    }

    @Override
    public void renderHeaderContainer(int serving, int preparationMinutes, int cookingMinutes
            , String recipeName) {
        containerHeader.setVisibility(View.VISIBLE);
        tvRecipeName.setText(recipeName);
        tvServingCount.setText(getString(R.string.detail_serving_format, serving));
        tvPreparationTime.setText(getString(R.string.detail_time_format, preparationMinutes));
        tvCookingTime.setText(getString(R.string.detail_time_format, cookingMinutes));
    }

    @Override
    public void renderIngredientCard(List<Ingredient> ingredients) {
        final List<String> formattedInstructions = new ArrayList<>(ingredients.size());
        int number = 1;
        for (Ingredient ingredient : ingredients) {
            formattedInstructions.add(getString(R.string.general_number_text_unit_format,
                    number, ingredient.getName(), ingredient.getAmount(), ingredient.getUnit()));
            number++;
        }
        cardIngredient.setVisibility(View.VISIBLE);
        rvIngredient.setNestedScrollingEnabled(false);
        rvIngredient.setLayoutManager(new LinearLayoutManager(getActivity(),
                LinearLayoutManager.VERTICAL, false));
        final TextViewRecyclerAdapter adapter = new TextViewRecyclerAdapter(getActivity(),
                formattedInstructions);
        rvIngredient.setAdapter(adapter);
    }

    @Override
    public void renderInstructionCard(List<Instruction> instructions) {
        final List<String> formattedInstructions = new ArrayList<>(instructions.size());
        for (Instruction instruction : instructions) {
            formattedInstructions.add(getString(R.string.general_number_text_format,
                    instruction.getNumber(), instruction.getStep()));
        }
        cardInstruction.setVisibility(View.VISIBLE);
        rvInstruction.setNestedScrollingEnabled(false);
        rvInstruction.setLayoutManager(new LinearLayoutManager(getActivity(),
                LinearLayoutManager.VERTICAL, false));
        final TextViewRecyclerAdapter adapter = new TextViewRecyclerAdapter(getActivity(),
                formattedInstructions);
        rvInstruction.setAdapter(adapter);
    }

    @Override
    public void setupSwipeRefreshLayout() {
        swipeRefreshLayout.setColorSchemeColors(ContextCompat.getColor(getActivity(), R.color.red));
        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                mPresenter.getRecipeDetailInformation(getArguments().getString(BUNDLE_RECIPE_ID));
            }
        });
    }

    @Override
    public void setupAdView() {
        adView.loadAd(new AdRequest.Builder().build());
    }

    @Override
    public void createShareIntent(String url, String recipeName) {
        final String textToShare = getString(R.string.detail_intent_share, recipeName, url);
        Intent shareIntent = new Intent(Intent.ACTION_SEND);
        shareIntent.setType("text/plain");
        shareIntent.putExtra(Intent.EXTRA_TEXT, textToShare);
        startActivity(Intent.createChooser(shareIntent, getString(R.string.general_sharechooser)));
    }

    @Override
    public void navigateToWebViewActivity(String url) {
        mFragmentListener.navigateToWebViewActivity(url);
    }

    interface FragmentListener {
        void navigateToWebViewActivity(String url);
    }
}

File: app/src/main/java/starbright/com/projectegg/features/detail/RecipeDetailFragment.kt
/**
 * Created by Andreas on 11/8/2019.
 */

package starbright.com.projectegg.features.detail

import android.support.v4.app.Fragment
import starbright.com.projectegg.data.model.Ingredient
import starbright.com.projectegg.data.model.Instruction

class RecipeDetailFragment : Fragment(), RecipeDetailContract.View {
    override fun showProgressBar() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun hideProgressBar() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun hideScrollContainer() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun showScrollContainer() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun hideEmptyStateTextView() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun renderErrorStateTextView(errorMessage: String) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun renderEmptyStateTextView() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun renderBannerFoodImage(imageURL: String) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun renderHeaderContainer(serving: Int, preparationMinutes: Int, cookingMinutes: Int, recipeName: String) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun renderIngredientCard(ingredients: MutableList<Ingredient>) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun renderInstructionCard(instructions: MutableList<Instruction>) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun setupSwipeRefreshLayout() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun createShareIntent(url: String, recipeName: String) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun navigateToWebViewActivity(url: String) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

//    @Inject
//    lateinit var schedulerProvider: SchedulerProviderContract
//    @Inject
//    lateinit var repository: AppRepository
//
//    private lateinit var mPresenter: RecipeDetailContract.Presenter
//    private var mFragmentListener: FragmentListener? = null
//
//    override fun onAttach(context: Context?) {
////        MyApp.getAppComponent().inject(this)
//        super.onAttach(context)
//        mFragmentListener = context as FragmentListener?
//    }
//
//    override fun onCreate(savedInstanceState: Bundle?) {
//        super.onCreate(savedInstanceState)
//        setHasOptionsMenu(true)
//        RecipeDetailPresenter(repository, this, schedulerProvider)
//    }
//
//    override fun onCreateView(inflater: LayoutInflater,
//                              container: ViewGroup?,
//                              savedInstanceState: Bundle?): View? {
//        return inflater.inflate(R.layout.fragment_recipe_detail, container, false)
//    }
//
//    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
//        super.onViewCreated(view, savedInstanceState)
//        mPresenter.apply {
//            setRecipeId(arguments!!.getString(BUNDLE_RECIPE_ID)!!)
//            start()
//        }
//    }
//
//    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater) {
//        inflater.inflate(R.menu.recipe_detail_menu, menu)
//    }
//
//    override fun onOptionsItemSelected(item: MenuItem): Boolean {
//        when (item.itemId) {
//            R.id.menu_share -> {
//                mPresenter.handleShareMenuClicked()
//                return true
//            }
//            R.id.menu_webview -> {
//                mPresenter.handleWebViewMenuClicked()
//                return true
//            }
//        }
//        return false
//    }
//
//    override fun setPresenter(presenter: RecipeDetailPresenter) {
//        mPresenter = presenter
//    }
//
//    override fun showProgressBar() {
//        swipe_refresh_container.isRefreshing = true
//    }
//
//    override fun hideProgressBar() {
//        swipe_refresh_container.isRefreshing = false
//    }
//
//    override fun hideScrollContainer() {
//        scroll_container.visibility = View.GONE
//    }
//
//    override fun showScrollContainer() {
//        scroll_container.visibility = View.VISIBLE
//    }
//
//    override fun hideEmptyStateTextView() {
//        tv_empty_text.visibility = View.GONE
//    }
//
//    override fun renderErrorStateTextView(errorMessage: String) {
//        tv_empty_text.run {
//            visibility = View.VISIBLE
//            text = errorMessage
//        }
//    }
//
//    override fun renderEmptyStateTextView() {
//        tv_empty_text.run {
//            visibility = View.VISIBLE
//            text = getString(R.string.detail_empty_label)
//        }
//    }
//
//    override fun renderBannerFoodImage(imageURL: String) {
//        img_banner_food.visibility = View.VISIBLE
//        GlideApp.with(activity!!)
//                .load(imageURL)
//                .diskCacheStrategy(DiskCacheStrategy.ALL)
//                .centerCrop()
//                .into(img_banner_food)
//    }
//
//    override fun renderHeaderContainer(serving: Int, preparationMinutes: Int, cookingMinutes: Int, recipeName: String) {
//        container_header.visibility = View.VISIBLE
//        tv_recipe_title.text = recipeName
//        tv_serving_count.text = getString(R.string.detail_serving_format, serving)
//        tv_preparation_time.text = getString(R.string.detail_time_format, preparationMinutes)
//        tv_cooking_time.text = getString(R.string.detail_time_format, cookingMinutes)
//    }
//
//    override fun renderIngredientCard(ingredients: MutableList<Ingredient>) {
//        val formattedInstructions = ArrayList<String>(ingredients.size)
//        var number = 1
//        for (ingredient in ingredients) {
//            formattedInstructions.add(getString(R.string.general_number_text_unit_format,
//                    number, ingredient.name, ingredient.amount, ingredient.unit))
//            number++
//        }
//        card_ingredient.visibility = View.VISIBLE
//        val adapter = TextViewRecyclerAdapter(activity,
//                formattedInstructions)
//        rv_ingredient.run {
//            isNestedScrollingEnabled = false
//            layoutManager = LinearLayoutManager(activity,
//                    LinearLayoutManager.VERTICAL, false)
//            this.adapter = adapter
//        }
//    }
//
//    override fun renderInstructionCard(instructions: MutableList<Instruction>) {
//        val formattedInstructions = ArrayList<String>(instructions.size)
//        for (instruction in instructions) {
//            formattedInstructions.add(getString(R.string.general_number_text_format,
//                    instruction.number, instruction.step))
//        }
//        card_instruction.visibility = View.VISIBLE
//        val adapter = TextViewRecyclerAdapter(activity,
//                formattedInstructions)
//        rv_instruction?.run {
//            isNestedScrollingEnabled = false
//            layoutManager = LinearLayoutManager(activity,
//                    LinearLayoutManager.VERTICAL, false)
//            this.adapter = adapter
//        }
//    }
//
//    override fun setupSwipeRefreshLayout() {
//        swipe_refresh_container.run {
//            setColorSchemeColors(ContextCompat.getColor(activity!!, R.color.red))
//            setOnRefreshListener { mPresenter.getRecipeDetailInformation(arguments!!.getString(BUNDLE_RECIPE_ID)!!) }
//        }
//    }
//
//    override fun createShareIntent(url: String, recipeName: String) {
//        val textToShare = getString(R.string.detail_intent_share, recipeName, url)
//        val shareIntent = Intent(Intent.ACTION_SEND)
//        shareIntent.type = "text/plain"
//        shareIntent.putExtra(Intent.EXTRA_TEXT, textToShare)
//        startActivity(Intent.createChooser(shareIntent, getString(R.string.general_sharechooser)))
//    }
//
//    override fun navigateToWebViewActivity(url: String) {
//        mFragmentListener?.navigateToWebViewActivity(url)
//    }
//
//    internal interface FragmentListener {
//        fun navigateToWebViewActivity(url: String)
//    }
//
//    companion object {
//
//        private const val BUNDLE_RECIPE_ID = "BUNDLE_RECIPE_ID"
//
//        fun newInstance(recipeId: String): RecipeDetailFragment {
//            val args = Bundle()
//            args.putString(BUNDLE_RECIPE_ID, recipeId)
//            val fragment = RecipeDetailFragment()
//            fragment.arguments = args
//            return fragment
//        }
//    }
}
